xwayland/xwm.c

```diff
@@ -0,0 +1,341 @@
+#include <stdlib.h>
+#include <xcb/xcb_event.h>
+#include <xcb/composite.h>
+#include "wlr/util/log.h"
+#include "wlr/types/wlr_surface.h"
+#include "wlr/xwayland.h"
+#include "xwm.h"
+
+/* General helpers */
+// TODO: replace this with hash table?
+static struct wlr_x11_window *lookup_window(struct wl_list *list, xcb_window_t window_id) {
+	struct wlr_x11_window *window;
+	wl_list_for_each(window, list, link) {
+		if (window->window_id == window_id) {
+			return window;
+		}
+	}
+	return NULL;
+}
+static struct wlr_x11_window *lookup_window_any(struct wlr_xwm *xwm, xcb_window_t window_id) {
+	struct wlr_x11_window *window;
+	if ((window = lookup_window(&xwm->xwayland->displayable_windows, window_id)) ||
+			(window = lookup_window(&xwm->unpaired_windows, window_id)) ||
+			(window = lookup_window(&xwm->new_windows, window_id))) {
```

Comment:
I'm not sure where these three lists should be, they all used to be in xwm but the compositor needed to access the first one.
Basically an wlr_x11_window is always in a list, either new_windows just after creation; for a very short time unpaired (depending on the order in which we get the wayland and the X11 notifications for surface creation) then all will be in displayable_windows